### PR TITLE
Read receipts for threads

### PR DESCRIPTION
**Pull request by gsouquet**
For https://github.com/vector-im/element-web/issues/23191
View full MSC details https://github.com/matrix-org/matrix-spec-proposals/pull/3771

This PR enables read receipts for threads.

If the server advertises support for `org.matrix.msc3771`, the JS SDK will comply and allow sending and receiving read receipts for threads. 

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Read receipts for threads ([\#2635](https://github.com/matrix-org/matrix-js-sdk/pull/2635)).

### CODE REVIEW

File: src/models/read-receipt.ts [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        };
+    };
+}
+
+const ReceiptPairRealIndex = 0;
+const ReceiptPairSyntheticIndex = 1;
+// We will only hold a synthetic receipt if we do not have a real receipt or the synthetic is newer.
+type Receipts = {
+    [receiptType: string]: {
+        [userId: string]: [WrappedReceipt, WrappedReceipt]; // Pair<real receipt, synthetic receipt> (both nullable)
```
Comment (kerryarchibald) [inline src/models/read-receipt.ts:70]: should this be `[userId: string]: [WrappedReceipt | null, WrappedReceipt | null]`?

File: src/models/read-receipt.ts [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+import { ReceiptType } from "../@types/read_receipts";
+import { EventTimelineSet, EventType, MatrixEvent } from "../matrix";
+import { ListenerMap, TypedEventEmitter } from "./typed-event-emitter";
+import * as utils from "../utils";
+
+export const MAIN_ROOM_TIMELINE = "main";
+
+export function synthesizeReceipt(userId: string, event: MatrixEvent, receiptType: ReceiptType): MatrixEvent {
+    // console.log("synthesizing receipt for "+event.getId());
```
Comment (kerryarchibald) [inline src/models/read-receipt.ts:22]: lost log

File: src/models/room.ts [hunk 3]
```diff
@@ -2461,6 +2409,37 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
...
+     * @param {MatrixEvent} event The m.receipt event.
+     * @param {Boolean} synthetic True if this event is implicit.
+     */
+    public addReceipt(event: MatrixEvent, synthetic = false): void {
+        const content = event.getContent<ReceiptContent>();
+        Object.keys(content).forEach((eventId: string) => {
+            Object.keys(content[eventId]).forEach((receiptType: ReceiptType) => {
+                Object.keys(content[eventId][receiptType]).forEach((userId: string) => {
+                    const receipt = content[eventId][receiptType][userId] as any;
+                    const receiptForMainTimeline = !receipt.threafd_id || receipt.thread_id === MAIN_ROOM_TIMELINE;
```
Comment (kerryarchibald) [inline src/models/room.ts:2423]: typo `threafd_id`?

File: src/models/room.ts [hunk 4]
```diff
@@ -2461,6 +2409,37 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
...
+     * Add a receipt event to the room.
+     * @param {MatrixEvent} event The m.receipt event.
+     * @param {Boolean} synthetic True if this event is implicit.
+     */
+    public addReceipt(event: MatrixEvent, synthetic = false): void {
+        const content = event.getContent<ReceiptContent>();
+        Object.keys(content).forEach((eventId: string) => {
+            Object.keys(content[eventId]).forEach((receiptType: ReceiptType) => {
+                Object.keys(content[eventId][receiptType]).forEach((userId: string) => {
+                    const receipt = content[eventId][receiptType][userId] as any;
```
Comment (kerryarchibald) [inline src/models/room.ts:2422]: Why any? Could you do some type assertion on this?

File: src/models/room.ts [hunk 5]
```diff
@@ -2461,6 +2409,37 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
...
+                    const receiptForMainTimeline = !receipt.threafd_id || receipt.thread_id === MAIN_ROOM_TIMELINE;
+                    const receiptDestination = receiptForMainTimeline
+                        ? this
+                        : this.threads.get(receipt.thread_id);
```
Comment (kerryarchibald) [inline src/models/room.ts:2426]: Could this be falsy?

**Review by kerryarchibald — COMMENTED** [review]
Looks good, I left a few nitpicks. Feel free to ignore the comments on code you just moved!

File: src/models/room.ts (hunk 4 above)
Comment (gsouquet) [reply src/models/room.ts:2422]: Yes, that's the remaining of copy pasting some other stuff!

File: src/models/room.ts (hunk 5 above)
Comment (gsouquet) [reply src/models/room.ts:2426]: Looks like it can, and the strict null check CI is complaining about it too!

File: src/models/room.ts [hunk 6]
```diff
@@ -2436,17 +2391,17 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
                 }
 
                 if (shouldLiveInThread) {
```
Comment (kerryarchibald) [inline src/models/room.ts:2393]: maybe the check for threadId should be in the if condition instead? Feels a bit code-smelly to be setting threadId to an empty string

File: src/models/room.ts [hunk 7]
```diff
@@ -2458,12 +2413,43 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
         const threadRoots = new Set<string>();
         for (const event of events) {
             if (event.isRelation(THREAD_RELATION_TYPE.name)) {
```
Comment (kerryarchibald) [inline src/models/room.ts:2415]: same?

**Review by kerryarchibald — APPROVED** [review]

File: src/models/room.ts (hunk 6 above)
Comment (gsouquet) [reply src/models/room.ts:2393]: Yes that's what I thought, but doing that would also prevent it from being added to the correct partition and would therefore not end up in any timeline. Which sounds unfortunate 🤔

**Pull request merged**